File: 백준/Platinum/14942. 개미/개미.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <numeric>

using namespace std;

constexpr int INF = 1e9;
using FNode = pair<int, int>;

int main()
{
    cin.tie(nullptr)->sync_with_stdio(false);

    int n; cin >> n;

    vector<int> EnergyofAnts(n + 1, 0);
    for (int i = 1; i <= n; ++i)
    {
        cin >> EnergyofAnts[i];
    }

    vector<vector<pair<int, int>>> Tree(n + 1);
    for (int i = 0; i < n - 1; ++i)
    {
        int a, b, c; cin >> a >> b >> c;

        Tree[a].emplace_back(b, c);
        Tree[b].emplace_back(a, c);
    }

    priority_queue<FNode, vector<FNode>, greater<>> PQ;
    vector<int> Distances(n + 1, INF);
    vector<int> Parents(n + 1, 0);

    PQ.emplace(0, 1);
    Distances[1] = 0;

    while (!PQ.empty())
    {
        const auto [du, u] = PQ.top();
        PQ.pop();

        if (Distances[u] < du) continue;

        for (const auto& [v, dv] : Tree[u])
        {
            if (du + dv < Distances[v])
            {
                Distances[v] = du + dv;
                Parents[v] = u;
                PQ.emplace(Distances[v], v);
            }
        }
    }

    vector<int> Answer(n + 1, 0);
    for (int i = 1; i <= n; ++i)
    {
        if (i == 0 || i == 1) continue;
        if (Tree[i].size() != 1) continue;

        int Node = i;
        deque<pair<int, int>> Paths;

        while (Node != 0)
        {
            Paths.emplace_back(Node, Distances[Node]);
            Node = Parents[Node];
        }

        while (!Paths.empty())
        {
            const auto [u, du] = Paths.front();
            if (Answer[u] != 0)
            {
                Paths.pop_front();
                continue;
            }

            int Lower = -1, Upper = Paths.size();
            while (Lower + 1 < Upper)
            {
                int Mid = (Lower + Upper) / 2;

                if (Paths[0].second - Paths[Mid].second <= EnergyofAnts[u])
                {
                    Lower = Mid;
                }
                else
                {
                    Upper = Mid;
                }
            }

            if (Lower == -1)
            {
                Answer[u] = Paths[0].first;
            }
            else
            {
                Answer[u] = Paths[Lower].first;
            }

            Paths.pop_front();
        }
    }

    for (int Node = 1; Node <= n; ++Node)
    {
        cout << Answer[Node] << "\n";
    }

    return 0;
}

File: 백준/Platinum/14942. 개미/개미.py
from heapq import *
from collections import deque


def binary_search(paths: deque, energy: int) -> int:
    lo, hi = -1, len(paths)
    while lo + 1 < hi:
        mid = (lo + hi) // 2
        if paths[0][1] - paths[mid][1] <= energy:
            lo = mid
        else:
            hi = mid
            
    return 0 if lo == -1 else lo
    

def dijkstra(heap: list, distances: list, parents: list, tree: list):
    heappush(heap, (0, 1))
    distances[1] = 0
    
    while heap:
        du, u = heappop(heap)
        
        if distances[u] < du:
            continue
        
        for v, dv in tree[u]:
            if du + dv < distances[v]:
                distances[v] = du + dv
                parents[v] = u
                heappush(heap, (distances[v], v))

    
def main():
    input = open(0).readline
    
    n = int(input())
    energy_of_ants = [0] + [int(input()) for _ in range(n)]
    tree = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        a, b, c = map(int, input().split())
        tree[a].append((b, c))
        tree[b].append((a, c))
                
    
    heap = []
    distances = [float('inf')] * (n + 1)
    parents = [0] * (n + 1)
    
    dijkstra(heap, distances, parents, tree)

    answer = [0] * (n + 1)
    for i in range(2, n + 1):
        if len(tree[i]) > 1:
            continue
        
        node = i
        paths = deque()
        while node != 0:
            paths.append((node, distances[node]))
            node = parents[node]
        
        while paths:
            u, du = paths[0]
            if answer[u] != 0:
                paths.popleft()
                continue
            
            answer[u] = paths[binary_search(paths, energy_of_ants[u])][0]
            paths.popleft()
    
    print('\n'.join(map(str, answer[1:])))


if __name__ == "__main__":
    main()